### PR TITLE
 fix(endpoints): scrape endpoints and not services

### PR DESCRIPTION
**Pull request by paologallinaharbur**
Fix https://github.com/newrelic/nri-prometheus/issues/27

The idea is that when a service is detected the underlying endpoints are fetched. For backward compatibility the metric directly coming from the service is kept
Screen Shot 2021-03-10 at 11 57 09 AM

Two new config options have been added `ScrapeServices (default true)` and `ScrapeEndpoints(default false)`
 
 After merging we should update documentation and the helm chart to ask for listing/watching permission on endpoints

### CODE REVIEW

**Comment by paologallinaharbur** [comment]
Currently after merging this we will be sending more data than before.

We should decide to filter out some targets, es:
```go
func (k *KubernetesTargetRetriever) GetTargets() ([]Target, error) {
	length := 0
	k.targets.Range(func(_, _ interface{}) bool {
		length++
		return true
	})
	targets := make([]Target, 0, length)
	k.targets.Range(func(_, y interface{}) bool {
		for _, t := range y.([]Target) {
			if t.Object.Kind == "service" && excludeServiceTargets {
				continue
			}
			if t.Object.Kind == "endpoints" && excludeEndpointsTargets {
				continue
			}
			targets = append(targets, t)
		}
		return true
	})

	return targets, nil
}
```

File: internal/pkg/endpoints/kubernetes.go
Comment (gsanchezgavier) [inline internal/pkg/endpoints/kubernetes.go:612]: we might need to cache the services and the endpoints related to avoid the calls to the API

**Review by gsanchezgavier — COMMENTED** [review]
I think we need  to have these behind a feature flag keeping the old behavior by default. Or have a mayor release. I would also propose to have this feature in beta for some time.  

Could you try the load test to se how is the result with the new feature?

File: cmd/k8s-target-retriever/main.go
Comment (davidbrota) [inline cmd/k8s-target-retriever/main.go:47]: why 7 and not 5? should it be made configurable?

Comment (paologallinaharbur) [reply cmd/k8s-target-retriever/main.go:47]: this is in the "test" main, it is not used by the real pomi

Comment (paologallinaharbur) [reply cmd/k8s-target-retriever/main.go:47]: it is used to test an debug the targets retrieved

**Comment by paologallinaharbur** [comment]
@gsanchezgavier I added two feature flags and given what we talked about offline in the past days, for now the option to scrape endpoints will be disabled by default.

I'll update the documentation and the HelmChart to further clarify this

**Review by gsanchezgavier — APPROVED** [review]
lgtm

**Comment by paologallinaharbur** [comment]
After merging and releasing we need to update documentation and expose the new config through the helm charts

File: cmd/nri-prometheus/config.go
Comment (roobre) [inline cmd/nri-prometheus/config.go:113]: I'd rename these without the `enable_` prefix

```suggestion
	viper.SetDefault("scrape_services", true)
	viper.SetDefault("scrape_endpoints", false)
```

File: deploy/nri-prometheus.tmpl.yaml
Comment (roobre) [inline deploy/nri-prometheus.tmpl.yaml:119]: Could we check this in the code? Since it is true by default, it is possible that some users set `enable_scrape_endpoints: true` but forget to disable the old one.

In fact, if the options are mutually exclusive, I'd strip `enable_scrape_services` entirely.

Comment (paologallinaharbur) [reply deploy/nri-prometheus.tmpl.yaml:119]: I am not totally sure, but I can imagine a possible use-case for which they actually want both.
Otherwise it would be quite obscure that enabling endpoints you lose services

File: cmd/nri-prometheus/config.go
Comment (paologallinaharbur) [reply cmd/nri-prometheus/config.go:113]: mmm, I added it since there was just above disale_... but I can change that :)

File: internal/pkg/endpoints/kubernetes.go
Comment (roobre) [inline internal/pkg/endpoints/kubernetes.go:630]:
```suggestion
		} else {
```

Comment (paologallinaharbur) [reply internal/pkg/endpoints/kubernetes.go:630]: didn't you saw "Please, do not try to reduce the amount of code below or simplify the conditionals."?

I'll update that xD

**Review by roobre — APPROVED** [review]

**Pull request merged**
